src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java

```diff
@@ -615,16 +606,20 @@ private String getRequest(String path) throws IOException, InterruptedException
         }
     }
 
-    private int getRequestStatus(String path) throws IOException {
+    private int headRequestStatus(String path) throws IOException, InterruptedException {
         HttpClient client = getHttpClient();
-        GetMethod httpget = new GetMethod(path);
+        HeadMethod httpHead = new HeadMethod(path);
         try {
-            executeMethod(client, httpget);
-            return httpget.getStatusCode();
+            executeMethod(client, httpHead);
+            while (httpHead.getStatusCode() == API_RATE_LIMIT_CODE) {
+                Thread.sleep(5000);
+                executeMethod(client, httpHead);
+            }
```

Comment:
:bug: Waiting forever here until the rate limit is ok to process seems wrong to me. I think failing with an exception (so the user can see it in the build/indexing log would be much better).

Reply:
Hmm, I agree with you, there should be some timeout (?). However,  this is the pattern among all http outgoing calls in existing code. for example, see: https://github.com/vivek/bitbucket-branch-source-plugin/blob/49e2d0791848ded3f5281279ef73bf7e7ebdf84b/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java#L586.

@stephenc how's GitHub branch source deals with api rate limit?

Reply:
So we should wait for ever, but check and respond to `Thread.interrupt()` and throw an `InterruptedException`

Reply:
Ok, I will adopt same approach here.

Reply:
And print some message in the build log? gh-branch-source does that IIRC

Reply:
@amuniz Not doing logging in this PR, needs further refactoring conflating this PR.

Reply:
Well, gh-branch-source is not waiting forever, it is just [throwing an exception](https://github.com/jenkinsci/github-branch-source-plugin/blob/ecfe2e072a56fa9b7e30183b21d526034e926fb7/src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java#L980).

Waiting there forever without any user visible effect is not correct IMO. What if there are other processes making requests to BB in parallel so the rate limit is continuously reached? Is this method going to hang forever? Better throw an exception and let it go to the build/indexing log so the user is aware of the problem.

Sorry, I'm keeping my :bug: here.

Reply:
@amuniz we are at a disagreement then. IMHO an SCMSource scan should not fail just because of a timeout, it should block until finished - while being responsive to interrupts. If we want to add a timeout to scanning, we can do that at the branch-api level by having a background thread that does the interrupt at the appropriate time.

As such I view this code as correct and I would :bug: if it threw a timeout. We had a lot of issues previously with branches getting lost because of a failure to behave correctly in this regard

Reply:
> We had a lot of issues previously with branches getting lost because of a failure to behave correctly in this regard

Wasn't it a bug in branch-api? If the retrieve process fails with an exception then the items clean up should not run? Otherwise there is a bug in gh-branch-source (which is throwing an exception).